File: setup/myapp/app/javascript/channels/matchmaking_channel.js
import consumer from "./consumer"
import $ from 'jquery'
import Helper from "../packs/Helper"

const Matchmaking = {}

$(function () {

  class MatchmakingChannel {

    constructor() {
      this.cable = null;
    }

    connect(user, callback, view, action = 'quick_game', peer = undefined) {

      if (this.cable) {
        return ;
      }
      const self = this;
      this.cable = consumer.subscriptions.create({
        channel: "MatchmakingChannel",
        id: user,
        action: action,
        peer: peer,
      },
      {
        connected() {
          // Called when the subscription is ready for use on the server
          if (action === 'quick_game' || action === 'ranked_game' || action === 'tournament_game' || action === 'war_game')
            this.perform(action)
          else if (action === 'wait_peer')
            this.perform(action, {"peer": peer, "from": this.id})
          else if (action === 'accept_peer')
            this.perform(action, {"peer": peer, "from": this.id})
          },

        disconnected() {
          // Called when the subscription has been terminated by the server
          self.disconnect()
        },

        received(data) {
          // Called when there's incoming data on the websocket for this channel
          callback.bind(view)(data)
        }
      });
    }

    disconnect() {
      if (this.cable) {
          consumer.subscriptions.remove(this.cable);
          this.cable = null;
      }
    }

  }

  Matchmaking.channel = new MatchmakingChannel();

});

export default Matchmaking;

File: setup/myapp/app/channels/matchmaking_channel.rb
class MatchmakingChannel < ApplicationCable::Channel
  def subscribed
    stream_from "Matchmaking_#{params[:id]}"
  end

  def unsubscribed
    Matchmaking.destroy_by(user_id: current_user.id)
  end

  def quick_game
    QuickGameJob.perform_later(current_user)
  end

  def ranked_game
    RankedGameJob.perform_later(current_user)
  end

  def tournament_game
    user = User.find_by(id: current_user.id)
    if user.intournament == true
      TournamentGameJob.perform_later(current_user)
    else
      ActionCable.server.broadcast( "Matchmaking_#{user.id}", { action: 'not_tournament' })
    end
  end

  def war_game
    user = User.find_by(id: current_user.id)
    if user.guild_id
      guild = Guild.find_by(id: user.guild_id)
      if guild.inwar == true && guild.war_playing == false
        guild.war_playing = true
        guild.save
        war = War.find_by(id: guild.war_id)
        WarGameJob.perform_later(user, war)
      elsif guild.war_playing == true
        ActionCable.server.broadcast( "Matchmaking_#{user.id}", { action: 'try_later' })
      else
        ActionCable.server.broadcast( "Matchmaking_#{user.id}", { action: 'not_war' })
      end
    else
      ActionCable.server.broadcast( "Matchmaking_#{user.id}", { action: 'not_war' })
    end
  end

  def wait_peer(data)
    puts data["peer"]
    puts "finding someone to dance with"
    user = User.find_by(id: data["peer"].to_i)
    user.send_notification('notification', 'challenge', current_user.nickname, {"from": current_user.id})
    ChallengeJob.perform_later(current_user, User.find_by(id: data["peer"].to_i))
  end

  def accept_peer(data)
    puts data["peer"]
    puts "accepting someone to dance with"
    user = User.find_by(id: data["peer"].to_i)
    #user.send_notification('notification', 'challenge', current_user.nickname, {"from": current_user.id})
    ChallengeJob.perform_later(current_user, User.find_by(id: data["peer"].to_i))
  end

  def miss_match
    user = User.find_by(id: current_user.id)
    guild = Guild.find_by(id: user.guild_id)
    war = War.find_by(id: guild.war_id)
    guild.warvictories += 1
    guild.missed_matches += 1
    guild.war_playing = false
    guild.save
    if (guild.missed_matches >= war.missed_matches)
      WarTimeOffJob.perform_now(war)
    end
  end

end
